### PR TITLE
Seq2seq trainer

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR graduates `Seq2SeqTrainer` and moves it inside the transformers library. By doing so, it moves some of the features of the `Seq2SeqTrainer` inside the main `Trainer` and leaves some in the subclass. More precisely, the following features will be available in the general Trainer:
- label smoothing is passed to the main `Trainer` (so it can be used in all classification problems), with an easier API, bug fixes (the current implementation did not work with -100 has an ignore index for the loss, now it does; also the current implementation returns a loss that is not averaged thus being too big, see below)
- the ability to pick any scheduler
- the ability to use Adafactor instead of AdamW

The sortish-sampling and predict with generate are left in the subclass.

There are also a few breaking changes in the `Seq2SeqTrainer` API to make its init match the one of `Trainer` mainly the init does not take a `config` and a `DataArguments`, instead:
- the token IDs are taken from the tokenizer
- the arguments for generation are passed to `evaluate` or `predict`
- the `ignore_pad_token_for_loss` is passed in the init but is deprecated since it should be removed once BART and subclasses use -100.

About label smoothing and the mean vs sum. The current implementation takes the sum over the batch size and sequence length (only counting tokens that have a label != -100). This gives something that does not have the same scale as the usual cross entropy loss (which is the mean on the same dimensions) thus would require a special learning rate to be useful. With the fix, the label-smoothed loss had the same scale as the non-smoothed loss, which means the same command with the same learning rate should produce comparable results.

### CODE REVIEW

**Comment by sgugger** [comment]
Oh and this PR does not delete the old `Seq2SeqTrainer` just yet since I'm planning to merge it just before my vacation. That way if something goes horribly wrong, people can still use the old `Seq2SeqTrainer`. The plan would be to delete it in January after the new `Seq2SeqTrainer` has been tested.

**Review by stas00 — APPROVED** [review]
The porting part looks good with a few suggestions. I re-run tests - found an issue pre-this PR. I will post it separately.

I haven't delved into label smoothing yes, so I've skipped over those parts' logic.

File: examples/seq2seq/finetune_trainer.py
Comment (stas00) [reply examples/seq2seq/finetune_trainer.py:336]: Also it's somewhat confusing that generate/config uses `max_length` attribute but trainer `max_target_length` - if we are creating new args, why not match the core naming?

(and there is no min_length override, but that's another story)

Comment (patil-suraj) [reply examples/seq2seq/finetune_trainer.py:336]: for s2s, `max_target_length` makes sense IMO since here we have two seqs source seq, and target seq. Just `max_length` might create some confusion.

Comment (stas00) [reply examples/seq2seq/finetune_trainer.py:336]: Just to clarify I'm talking about the arg name to `predict` which then calls `generate`.

In no way I was proposing to change the cl args. 

Are we on the same page?

File: src/transformers/trainer_seq2seq.py
Comment (patil-suraj) [inline src/transformers/trainer_seq2seq.py:41]: BART now works with -100, the issue was with label smoothing (which you fixed) and the `shift_tokens_right` function, which is fixed in #9135

**Review by patil-suraj — APPROVED** [review]
This looks good to me, thanks Sylvain!

File: examples/seq2seq/finetune_trainer.py
Comment (patil-suraj) [reply examples/seq2seq/finetune_trainer.py:336]: Aah, I see, sorry, I misread it.

Comment (patrickvonplaten) [reply examples/seq2seq/finetune_trainer.py:336]: I agree with @stas00 here, I think it's cleaner if the arg is called `max_length`. `max_length` in `generate()` always corresponds to the max_length that is predicted which would never correspond to "`max_source_length`". Since `evaluate` and `predict` are very closely tied to `generate()`, I think it's nice if this is a bit reflected in the function's signature by calling the arg `max_length`.

Comment (patrickvonplaten) [reply examples/seq2seq/finetune_trainer.py:336]: Also do we really need `val_max_length` and `test_max_length`? -> Does it make much sense to have different values for those two? I can see why one would have `data_args.val_beams` smaller than `data_args.eval_beams`, but for `max_length` as well?

File: src/transformers/trainer_pt_utils.py
Comment (patrickvonplaten) [inline src/transformers/trainer_pt_utils.py:369]:
```suggestion
    Adds label-smoothing on a pre-computed output from a Transformers model.
```

File: src/transformers/trainer_pt_utils.py
Comment (patrickvonplaten) [inline src/transformers/trainer_pt_utils.py:375]: (question) Not sure if the `Set to :obj:`None` to ignore none of the indices` is really required. For me the default case should be `-100` but if the user does not want to ignore any indices, there should simply be no -100 in the `labels` no? Or is there a case where one would have -100 in the labels **and** wants to apply `label_smoothing` on those?

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [reply src/transformers/trainer_seq2seq.py:41]: From the `src/transformers` side we can completely remove the deprecation warning and `ignore_pad_token_for_loss` arg IMO. On the other hand, it seemed like `ignore_pad_token_for_loss` was the default case for all training code in `examples/seq2seq` -> so might be good to keep it as it is now for some time

File: examples/seq2seq/finetune_trainer.py
Comment (patil-suraj) [reply examples/seq2seq/finetune_trainer.py:336]: It was added for convenience, but not super important, it makes sense to keep just one `max_target_length`

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:92]:
```suggestion
            ignore_keys (:obj:`List[str]`, `optional`):
```

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:96]:
```suggestion
                An optional prefix to be used as the metrics key prefix. For example the metrics ``"bleu"`` will be named
```

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:97]:
```suggestion
                "eval_bleu" if the prefix is ``"eval"`` (default)
```

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:76]: would also prefer the arg to be name `max_length` instead of `max_target_length`

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:117]:
```suggestion
        max_length: Optional[int] = None,
```

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:130]:
```suggestion
            ignore_keys (:obj:`List[str]`, `optional`):
```

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:135]:
```suggestion
                "eval_bleu" if the prefix is ``"eval"`` (default)
```

File: src/transformers/trainer_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/trainer_seq2seq.py:144]:
```suggestion
            If your predictions or labels have different sequence lengths (for instance because you're doing dynamic
```

File: src/transformers/training_args_seq2seq.py
Comment (patrickvonplaten) [inline src/transformers/training_args_seq2seq.py:39]:
```suggestion
    sortish_sampler: bool = field(default=False, metadata={"help": "Whether to use SortishSampler or not."})
```

File: tests/test_trainer_utils.py
Comment (patrickvonplaten) [inline tests/test_trainer_utils.py:63]: nice test!

**Review by patrickvonplaten — APPROVED** [review]
Great! Mostly nits from my side. Two things, I think are worth considering:

- change naming from `max_target_length` -> `max_length`. Think it's clearer this way that the args of `predict` and `eval` correspond 1-to-1 to the `max_length` of `generate()`
- since Seq2SeqTrainer is in `src/transformers` now , it might be worth to also add a test for it in `tests/test_trainer.py` IMO

File: examples/seq2seq/finetune_trainer.py
Comment (sgugger) [reply examples/seq2seq/finetune_trainer.py:336]: I haven't touched anything on the script except using `test_max_length` here, which was never used before. I thought it was just a mistake. Since there is a lot of debate on this and I'm not sure I understand what you want me to do, I will revert that and leave it as it was before. Then we can do some renaming in a separate PR (or please be clearer on what you want me to do :-) )

File: src/transformers/trainer_seq2seq.py
Comment (sgugger) [reply src/transformers/trainer_seq2seq.py:41]: Will adapt the TODO and the warning then, to check all mentions in the examples. Once we're sure it's not used anywhere, will remove the argument. but let's try to have it removed before the next release, so it's never a real "breaking change" to remove that arg.

File: src/transformers/trainer_pt_utils.py
Comment (sgugger) [reply src/transformers/trainer_pt_utils.py:375]: Yeah, the docstring is just badly worded.

File: examples/seq2seq/finetune_trainer.py
Comment (sgugger) [reply examples/seq2seq/finetune_trainer.py:336]: Oh I think I get it, so all `max_target_length` -> `max_length` and use one `eval_max_length`, like `eval_num_beams`. Will push that.

File: src/transformers/trainer_seq2seq.py
Comment (sgugger) [reply src/transformers/trainer_seq2seq.py:41]: On second thoughts, will have to change all scripts for the `max_target_length` -> `max_length` so I'll remove those `ignore_pad_token_for_loss` in passing. (just realized it is never actually used inside the code and the tests are still passing.)

File: examples/seq2seq/finetune_trainer.py
Comment (patil-suraj) [reply examples/seq2seq/finetune_trainer.py:336]: exactly, sorry for the confusing comments ;)

File: examples/seq2seq/finetune_trainer.py
Comment (stas00) [inline examples/seq2seq/finetune_trainer.py:96]: Why was this done in the last second? This and all the renames below just broke all the scripts that were relying on --max_target_length --val_max_target_length args.

Comment (stas00) [reply examples/seq2seq/finetune_trainer.py:96]: The discussion and the agreement as how I understood it was to rename only the new arguments you added and leaving everything else as before. We need source and target for datasets.

**Comment by stas00** [comment]
>     * change naming from `max_target_length` -> `max_length`. Think it's clearer this way that the args of `predict` and `eval` correspond 1-to-1 to the `max_length` of `generate()`

@patrickvonplaten - did you mean to suggest to change the new arguments to evaluate/predict that this PR adds or to rename `--max_target_length`, `--val_max_target_length` cl args?

**Comment by patrickvonplaten** [comment]
> suggest

I was more referring to the args of the functions, but more generally I think it would actually better if there would be only one `max_length` in the data_args - so leave the `max_length` that we have now and completely remove the `source_max_length` argument. IMO, `source_max_length` should be fully defined by the tokenizer of the model. I don't really see a need to let the user define the maximum input length, but this is probably better to be done in a separate PR. What is your opinion here @stas00 @sgugger @patil-suraj ?

File: examples/seq2seq/finetune_trainer.py
Comment (patrickvonplaten) [reply examples/seq2seq/finetune_trainer.py:96]: Sorry, I should have been a bit more careful at my second review of the PR...I'd probably prefer replacing all `--val_max_target_length` and `test_max_target_length` with `--max_length`, but I'm also fine with reverting the change on `DataTrainingArguments`

Comment (stas00) [reply examples/seq2seq/finetune_trainer.py:96]: Thank you for clarifying, so indeed you didn't not suggest to change the cl args as I read it.

I guess @sgugger read your mind then - I have no problem with that other than needed a deprecation cycle for such changes because it really interferes with my work at integrating / benchmarking / testing deepspeed and fairscale.

Then if we are now switching to refactoring the cl args - there should be no `eval_max_length` either, no?

Comment (patrickvonplaten) [reply examples/seq2seq/finetune_trainer.py:96]: IMO, one `max_length` should be enough. It does not make much sense to tokenizer longer than the model is able to generate and the other way around would result in errors. So IMO the best approach is to have one `max_length` referring to the maximum length that the model should be able to generate. 

The question then is whether we need a `max_source_length` or `max_seq_length` as it's called in `run_mlm.py` or not. I think you guys have more experience here.

**Comment by patil-suraj** [comment]
I think it's better to keep `max_source_length`, `max_target_length` since in some cases the input length could be way shorter than the tokenizer or model's max length and these two could be used to truncate the text .

We can get rid of `val_max_target_length` and `test_max_target_length` in `DataTrainingArguments`, since in almost all scripts we are using the same length for all three arguments (`max_target_length`,  `val_max_target_length`, `test_max_target_length`). Then we could pass the same `max_target_length`  to both `evaluate` and `predict` methods.

sorry about the miscommunication.

**Comment by stas00** [comment]
I don't know whether this is a good time, but should we add `min_length` as well here while this part of the API is being redesigned? Surely if generate has `min_length` someone might need to redefine it too? But I'm totally fine to deferring this until and if someone asks for it.

**Comment by stas00** [comment]
The best way to decide about the naming is to show a few use cases - and then it's clear whether these are needed or not.

Please don't forget that Sam or whoever added those in first place had a good reason for it, so it'd be hard to make a decision in the void.

**Pull request merged**
